### ISSUE TITLE
look_for_keys overwrites key_filename somehow.

### ISSUE THREAD

**Issue by dzatoah** [paramiko/paramiko#2112]
Debian package version: (python3-paramiko) 2.7.2-1

I've encountered an issue where providing a private key with `key_filename` is not enough to make paramiko use the key.
I also have to set `look_for_keys` to False. (It is true by default)

**Comment by bskinn** [comment]
Thanks for reporting this -- can you provide a code sample with a minimal reproduction of the problem, with debug output if possible?

**Comment by jun66j5** [comment]
`SSHClient.connect()` method tries to authenticate using a private key given for `key_filename` parameter. When the publickey authentication is failed, the `connect()` tries to authenticate using each private key in `~/.ssh` directory because `look_for_keys` is `True` by default.

I consider the `look_for_keys=True` is reasonable when `key_filename` parameter is not given but the user might not expect `look_for_keys=True` when `key_filename` parameter is given.

E.g. OpenSSH client, when a private key is given for `-i` option and fails to authenticate, doesn't try to authenticate with private keys in `~/.ssh` directory.

**Comment by dzatoah** [comment]
Yep thats exactly the problem. I found it while playing with `pyhoca-cli` (which uses python-x2go, which uses paramiko)
I fixed it temporarily by setting `look_for_keys=False` everytime I pass `key_filename`. But my commit had to be reverted since it breaks ssh-agent based authentication..
